# csc-gatk4-somatic-workflows
Introduction and Workflows for Somatic short variant discovery (SNVs + Indels) with GATK4 on CSC cPouta VM

This document offering the step-by-step introduction about calling somatic short variants with GATK4.1.2.0 Docker images, oncotator:1.9.9.0 Docker images (if needed) and Cromwell version 36 on CSC cPouta VM.
It mainly contains two wdl scripts and matched input json files: Mutect2.wdl and Mutect2_pon.wdl(optional).  
The main steps´description can be checked from [GATK Best Practice](https://gatkforums.broadinstitute.org/gatk/discussion/11146/somatic-short-variant-discovery-snvs-indels) and its matched [Github Page](https://github.com/gatk-workflows/gatk4-somatic-snvs-indels).  
In this [CSC Github Page](https://github.com/github-fish/csc-gatk4-somatic-workflows), we offers workflow scripts which suitable for running on cPouta VM and [testing results](https://github.com/github-fish/csc-gatk4-somatic-workflows/tree/master/TestingResult). Tested on cPouta VM, 24 cores, 117.2GB RAM.
## Mutect2
This WDL workflow runs GATK4 Mutect 2 on a single tumor-normal pair or on a single tumor sample, and performs additional filtering and functional annotation tasks.
***  
**Basic inputs**: Tumor bam and index file. Normal bam and index file.  
**Basic outputs**: Unfiltered vcf and index file. Filtered vcf and index file. Contamination.table. Mutect_stats:merged.stats. Filtering_stats: filtering.stats. Maf_segments: segments.table. [Basic Output](https://github.com/github-fish/csc-gatk4-somatic-workflows/blob/master/TestingResult/m2_v36_no_obFiler_no_oncotator_Aug27).  
**Full outputs**(need more input files which can be checked in mutect2.exome.inputs.json): [Full Output](https://github.com/github-fish/csc-gatk4-somatic-workflows/blob/master/TestingResult/m2_v36_full_Aug28).
### Tool Prerequisites
1. virtul machine (VM) on CSC cPouta
2. docker-ce
3. Git
### Running Instructions
1. Setup your working directory and make a directory to store your input files.
``` bash
mkdir gatk4-somatic-snvs-indels
cd gatk4-somatic-snvs-indels
mkdir inputs
```
2. Git clone needed github repo.
``` bash
git clone git://github.com/github-fish/csc-gatk4-somatic-workflows
```
3. Download Cromwell, the java excutable that will run the WDL.
``` bash
wget https://github.com/broadinstitute/cromwell/releases/download/36/cromwell-36.jar
```
4. Download need input files which are specified in json files. Store them in inputs directory.  
``` bash
cd inputs # Then download needed inputs file.
```
[CSC Bucket](https://object.pouta.csc.fi/swift/v1/AUTH_319722507e614d03b5b771f50da47361/Somatic_SNVs_Indels/), [Google Bucket1](https://console.cloud.google.com/storage/browser/gatk-best-practices/somatic-b37) and [Google Bucket2](https://console.cloud.google.com/storage/browser/broad-public-datasets/funcotator). CSC Bucket contains all needed testing inputs. Google Bucket 2 contains inputs for Funcotator, Google Bucket 1 contains all needed other inputs. Basically, CSC Bucket = Google Bucket 1 + 2.  

5. Edit the json file so that all file paths are replaced by your VM local file paths. And, this mutect2.exome.inputs.json file contains full functions of Mutect2.wdl. **Notice**: If you only want to get the VCF file and its index with primary filtering applied, just use mutect2.exome.inputs.basic.json file.
6. Change back to the main working dirctory.
``` bash
cd ../
```
At this point your directory structure should look like this
``` bash
[cloud-user@t gatk4-somatic-snvs-indels]$ tree
.
├── cromwell-36.jar
├── csc-gatk4-somatic-workflows
│   ├── mutect2.exome.inputs.bamout.json
│   ├── mutect2.exome.inputs.basic.json
│   ├── mutect2.exome.inputs.funcotator.json
│   ├── mutect2.exome.inputs.json
│   ├── mutect2.exome.inputs.ob_filer.json
│   ├── mutect2.exome.inputs.oncotator.json
│   ├── mutect2.wdl
│   ├── README.md
│   └── TestingResult
│       ├── m2_v36_bamout_Aug27
│       ├── m2_v36_full_Aug28
│       ├── m2_v36_funcotator_Aug28
│       ├── m2_v36_no_obFiler_no_oncotator_Aug27
│       ├── m2_v36_ob_filter_Aug27
│       ├── m2_v36_oncotator_27Aug
│       └── README.md
└── inputs
```
7. Execute your workflow
``` bash
sudo systemctl restart docker
sudo time java -jar cromwell-36.jar run ./csc-gatk4-somatic-workflows/mutect2.wdl -i ./csc-gatk4-somatic-workflows/mutect2.exome.inputs.json # "time" is optional for monitoring. According your needs, maybe mutect2.exome.inputs.json will be others, such as: mutect2.exome.inputs.basic.json.
```  

8. Expected result. While the workflow is running cromwell will print out logs to your screen. Once it completes it will print out a message indicating the run was successful. Also, it will print out the location of the output files that was generated by your workflow. [Looks Like This](https://github.com/github-fish/csc-gatk4-somatic-workflows/blob/master/TestingResult/m2_v36_full_Aug28).  
It will also generates two directories: cromwell-executions and cromwell-workflow-logs.
``` bash
[cloud-user@t gatk4-somatic-snvs-indels]$ ls
cromwell-executions  cromwell-workflow-logs  cromwell-36.jar csc-gatk4-somatic-workflows inputs
```
## Testing Result  
Pipelines were tested on cPouta VM (24 cores, 117.2 GB RAM) with small sized HCC1143 bam and matched normal bam file. Reference genome is b37. Other testing files/parameters are listed in this [json file](https://github.com/github-fish/csc-gatk4-somatic-workflows/blob/master/mutect2.exome.inputs.json). More detailed info and TerminalSavedOutput can be found in [TestingResult Folder](https://github.com/github-fish/csc-gatk4-somatic-workflows/tree/master/TestingResult). **Notice**: As we mentioned at 5, if you don´t want to generate all outputs, you don´t need to offer all inputs specified in this file. You can choose right json file for your target.
## Mutect2_Pon.wdl: Create Panel of Nomals (PON) for Mutect2 calling short varinats 
## Frequently fault report and solutions
## Reference
